### ISSUE TITLE
(Menu) Add Copy/Paste Support to On-Screen Keyboard

### ISSUE THREAD

**Issue by RobLoach** [libretro/RetroArch#10244]
Coming from https://github.com/libretro/libretro-database/issues/1016 , it would be great to be able to copy/paste text into the on-screen keyboard text input screen. This would particularly apply nicely to codes...

Brought up by @candrapersada

![75594850-d27d4d00-5abc-11ea-8f92-d1b01cbc6a9a](https://user-images.githubusercontent.com/25086/76276101-5f619c80-625b-11ea-9d20-31ef98cddd1f.png)
![75595107-a8785a80-5abd-11ea-91b1-4aaca466d42b](https://user-images.githubusercontent.com/25086/76276103-5ffa3300-625b-11ea-946b-5083b585e654.png)

**Comment by candrapersada** [comment]
![paste](https://user-images.githubusercontent.com/36833664/76279403-16443380-62c1-11ea-8e06-5f0178c81567.PNG)
not found paste in RetroArch

**Comment by Apocalypse612** [comment]
+1. Please implement. Typing ip addresses, passwords, codes, etc is such a hassle right now.

**Comment by JimR21** [comment]
Indeed, that would be very helpful. Writing my oversized, generated password manually in Retroarch is really a hassle.

**Comment by hizzlekizzle** [comment]
This would indeed be great, but I believe the issue is that clipboard access is a bit of a mess involving very platform-specific strategies for each one.

**Comment by phakic** [comment]
So I can suggest a workaround, at least for windows, if a little messy: use [phraseexpress](https://www.phraseexpress.com/), which is a text expanding program. Create a phrase which inserts clipboard content (for example, when I type ".v" it is replaced by whatever text I recently copied). Now, this is important, you need to modify a phraseexpress setting: under expert options, change the paste method to "paste text via keystroke simulation". Thus, when you type your phrase, it will be replaced by your clipboard content via keystroke simulation, which thus allows you to quickly insert copied text into retroarch. Hope this helps.

**Comment by eliashkan** [comment]
So as I understand this issue is here because the RetroArch in the PlayStore is exactly the same app as available for desktop systems? 🙄

**Comment by hizzlekizzle** [comment]
Yes unresolved at this time, and Android is one of the most complicated of cases because we don't/can't use the standard toolkit APIs.

**Comment by Cibomatto2002** [comment]
is there any updates on this ?

**Comment by RobLoach** [comment]
If there were updates, there would be a comment :wink: .... Click the "Subscribe" button on the right if you're looking to follow updates.

**Comment by rehhouari** [comment]
For Linux users:
1. install `xdotools` (debian/ubuntu: `sudo apt install xdotool`)
2. open the input on retroarch you want, for example twitch stream key
3. switch to a terminal
4. `sleep 5 && xdotool type 'text you want to paste' `
5. switch back to retroarch and make sure the input is focused
6. ???
7. profit

**Comment by cdrfun** [comment]
+1 for me. Adding my RetroAchievements account with a 30 char password is a pain.

**Comment by ShockwaveNN** [comment]
> +1 for me. Adding my RetroAchievements account with a 30 char password is a pain.

 
For me solutions for complicated RA password was direct edit config files, until this issue is resolved

**Comment by LibretroAdmin** [comment]
I could see drag and drop for Windows/Linux at least being feasible.

**Comment by cronyakatsuki** [comment]
> > +1 for me. Adding my RetroAchievements account with a 30 char password is a pain.
> 
> For me solutions for complicated RA password was direct edit config files, until this issue is resolved

Could you provide what exactly you need to edit?

**Comment by Diegovsky** [comment]
k, give me some pointers and I'll try sending a PR for this, only for desktop platforms to keep it simple for now

**Comment by zoltanvb** [comment]
Keyboard input is handled in input drivers: https://github.com/libretro/RetroArch/tree/master/input/drivers
Platform specific code is typically handled in frontend drivers: https://github.com/libretro/RetroArch/tree/master/frontend/drivers
As a first attempt, maybe `SDL_GetClipboardText` can be used in the SDL input driver without frontend involvement, just be aware that SDL input driver is only active when SDL video driver is also active: https://docs.libretro.com/guides/input-controller-drivers/

**Comment by Diegovsky** [comment]
Wow, that might be just what I need! Thank you :)

**Comment by Diegovsky** [comment]
@zoltanvb I wanted to discuss some RA implementation details to go about this. Would you prefer if I opened a draft PR and we can go over it there, or is there some other means like IRC/Matrix/Discord for faster communication?

**Comment by hizzlekizzle** [comment]
A combination of draft PR and [discord](https://discord.com/invite/VZ2b7wghxR) is usually a solid strategy. Once you're on the libretro discord server, head to #programming or #programming-frontend.

**Comment by ninjaloose** [comment]
Good to see some work on this, I just hit this issue myself trying to add a complex generated password for Retro Achievements. For anyone hunting for a way to put your password (or anything I suppose) in (on Windows at least), put in some unique dummy text like "tester1" and then you can open your "retroarch.cfg" that sits next to your Retroarch.exe and search for <your dummy text> ie "tester1", then you'll have found the field you need to change to the complicated one. This also lead me to an undisclosed exclusion requirement, you cannot have " as a character in your passwords due to how the password is stored in plaintext in the config file.

**Comment by cyberpaladin** [comment]
Wow, I'm kind of surprised this has been going on for years. Commenting to follow on, I also couldn't manually type my password, which is pretty complex. The virtual keyboard won't even show all characters and won't accept international keyboard input (some accentuation symbols).

**Comment by Diegovsky** [comment]
Aw shucks. I forgot all about this due to life and $work. I am not working on this at the moment. I tried reading the source code for a bit but I must confess: adding this seems like a lot of backend work. This is due to the fact that keys are added as-is to a buffer of current keys. It seems that there are no high level "Events" (delete, esc/cancel, etc), just key presses. This would somewhat require extensive changes under the hood, I think.

I didn't have much time to figure out what do to after I reached this conclusion. I was probably wrong, but nevertheless didn't have time ever since to revisit this again :/

**Comment by Hoff16** [comment]
Just as a note and not sure if this helps but on iOS recently the keyboard was changed to using the system keyboard instead of a unique RetroArch one. Could that be a way to get this solved or is that unrelated?

**Comment by Hoff16** [comment]
Here's the pull request for the keyboard change:

https://github.com/libretro/RetroArch/pull/18355

**Comment by warmenhoven** [comment]
The iOS native keyboard does already allow for pasting already, using Apple's UC (three finger tap above the keyboard brings up an edit menu). It is highly specific to iOS/tvOS.

**Comment by Hoff16** [comment]
> The iOS native keyboard does already allow for pasting already, using Apple's UC (three finger tap above the keyboard brings up an edit menu). It is highly specific to iOS/tvOS.

I can't seem to accomplish this. Where exactly do I need to tap?
![image](https://github.com/user-attachments/assets/96d59bc2-a7c3-4c5a-b8e9-eb3a8cc11011)

**Comment by warmenhoven** [comment]
I don't know exactly, above? Around where the prompt is worked for me. I don't think I've turned on any special accessibility features in the OS. I know retroarch isn't drawing this menu.

I'm attaching a video, usually it comes up for me first try but this time it didn't register right away so I had to tap again.

https://github.com/user-attachments/assets/83428a1e-7961-4fe2-907f-0bac70ff2cbb

**Comment by Hoff16** [comment]
> I don't know exactly, above? Around where the prompt is worked for me. I don't think I've turned on any special accessibility features in the OS. I know retroarch isn't drawing this menu.
> 
> I'm attaching a video, usually it comes up for me first try but this time it didn't register right away so I had to tap again.
> 
> https://github.com/user-attachments/assets/83428a1e-7961-4fe2-907f-0bac70ff2cbb

That's not the normal iOS paste dialog. It's something like this:

![image](https://github.com/user-attachments/assets/c89116a9-892d-4d17-b203-a932ff9004a8)

I'm tapping exactly where you are and I don't see any menu.

I'm also using glui if that makes a difference.

Thanks!

**Comment by warmenhoven** [comment]
You're right, it's not the "normal" iOS dialog, but it is one that is provided by iOS, retroarch doesn't build it itself or control the action that brings it up or what it looks like or what it does. The OS, iOS, decides all of that. And it works fine for me in glui:

https://github.com/user-attachments/assets/f030d66d-2ede-4c41-9712-13b298c0fbdb

I was tapping just above the Username dialog, quick three finger tap.

**Comment by warmenhoven** [comment]
To further demonstrate this is an OS-provided dialog, here it is in discord, when I three-finger-tap just above the keyboard (on the message):

https://github.com/user-attachments/assets/abe8d203-c691-460c-88a4-8fd9620a7829

**Comment by Hoff16** [comment]
> You're right, it's not the "normal" iOS dialog, but it is one that is provided by iOS, retroarch doesn't build it itself or control the action that brings it up or what it looks like or what it does. The OS, iOS, decides all of that. And it works fine for me in glui:
> 
> https://github.com/user-attachments/assets/f030d66d-2ede-4c41-9712-13b298c0fbdb
> 
> I was tapping just above the Username dialog, quick three finger tap.

I have never seen that menu and can't get three finger tap to display anything whatsoever. Let me dig into this.

**Comment by Hoff16** [comment]
> > You're right, it's not the "normal" iOS dialog, but it is one that is provided by iOS, retroarch doesn't build it itself or control the action that brings it up or what it looks like or what it does. The OS, iOS, decides all of that. And it works fine for me in glui:
> > 
> > https://github.com/user-attachments/assets/f030d66d-2ede-4c41-9712-13b298c0fbdb
> > 
> > I was tapping just above the Username dialog, quick three finger tap.
> 
> I have never seen that menu and can't get three finger tap to display anything whatsoever. Let me dig into this.

Nevermind, finally got it. I always just long-pressed to get a similar menu to appear but the three-finger tap does work! It's a quick three horizontal finger tap!

Thanks!
